Cookies & Refresh token (#14)
rsksmart/rif-identity-docs: ssi/specs/did-auth.markdown

layout: default
---

> # DID authentication - a challenge–response authentication model based on DIDs

Reviewer: ```suggestion
## DID authentication - a challenge–response authentication model based on DIDs
```

A challenge–response authentication is a family of protocols in which one party presents a question ("challenge") and another party must provide a valid answer ("response") to be authenticated. {% include ref.html id="10" %}

> # Table of contents

Reviewer: ```suggestion
### Table of contents
```

- [State of the art](#state-of-the-art)
- [Motivation](#motivation)
- [Endpoints](#endpoints)
  - [Signup](#signup)
    - [Selective disclosure](#selective-disclosure)
  - [Login](#login)
  - [Logout](#logout)
- [Implementations](#implementations)
- [Extensions](#extensions)
- [Open work](#open-work)
- [Apendix](#appendix)
  - [How to calculate a deterministic challenge](#how-to-calculate-a-deterministic-challenge)
  - [How to send tokens](#how-to-send-tokens)

> # State of the art

Reviewer: ```suggestion
### State of the art
```

> Nowadays, authentication is handled in a centralized way. Most applications delegate the process of authenticating users to third-party services ([OIDC](https://openid.net/connect/)), and the authenticity of that information relies on those services providers. The problem with this type of authentication is that the user's information is not controlled by her, the third-party services are those who control it. Today's decentralized applications cannot rely on this authentication, since they must prove that the user is the one who controls their information and not a third party, that's why a new protocol is required.

Reviewer: I didn't understand this part 
"Today's decentralized applications cannot rely on this authentication, since they must prove that the user is the one who controls their information and not a third party"

# Motivation

> The motivation of this protocol is to provide a user centric authentication and registration mechanism to be used by backend services of the web 3.0. There are two main type of services: _open_ (public access) and _permissioned_ (the user must provide certain information to access it). This proposal allows an external service to prove that the user controls her did and, if necessary, it may ask the user to share some private information associated to that did in order to register the user in the service.

Reviewer: ```suggestion
The motivation of this protocol is to provide user centric registration and authentication mechanisms to be used by services of the web 3.0. We identify two main type of services: _open_ (public access) and _permissioned_ (the user must provide certain information to access it). This protocol allows an external service to prove that the user is in control of their did and, optionally, enables the application to ask the user to share specific private information associated to their DID, in order to register the them.
```

It is designed to:

- Register users by requesting them to share specific information that can be verified - this information can be used in business logic to grant or deny access
- Allow the user to opt-in or out to the information - it is a user-centric protocol, the user decides wether to share the information or not
- Authenticate a user by proving that she controls a specific did
- Provide an _access token_ to the user that can be reused during a certain amount of time - wallet systems usually request user action to sign messages, enabling reusing access token reduces the amount of signatures required and improves the user experience

This protocol is based on the following specifications:
- CHAP{% include ref.html id="11" %} authentication protocol
- OAuth 2.0 Authorization Framework{% include ref.html id="13" %}.
- Decentralized Identifiers (DIDs) v1.0 {% include ref.html id="2" %}
- W3C Verifiable Credentials model {% include ref.html id="7" %}
- [uPort selective disclosure implementation](https://developer.uport.me/flows/selectivedisclosure)
- W3C Verifiable Credentials JSON Schema Specification {% include ref.html id="12" %}

> It uses HTTPS as the message transport layer.

Reviewer: ```suggestion
Initially, it is design to use HTTPS as the message transport layer, but can be ported to a different one.
```

> # Endpoints

Reviewer: ```suggestion
```

> ## Signup

Reviewer: ```suggestion
### Signup
```

> This section is OPTIONAL, it depends on the service needings. Some services may not need to register users before letting them enter in.

Reviewer: ```suggestion
Services usually need to register users before letting them in. This means the service requires some specific information to be share by the user (e.g.: user's email).

> This part of the protocol is OPTIONAL, it depends on the service needs. Some services may not need to register users before letting them enter in.
```

> 1. _Client_ sends `POST /request-signup { did }` to _Service_, where `did` is _User_'s DID

Reviewer: ```suggestion
1. _Client_ sends `POST /request-signup { did }` to _Service_, where `did` is _User_'s DID
```

How will the apps know if they need to fetch this or `request-auth` directly? Or is it that they always fetch this and then they fallback into `/request-auth`?

Let's think of a public service that you know it is DID AUth compliant but you are not sure yet of what the registration flow is.

> 2. _Service_ creates a deterministic _challenge_ (see [How to calculate a deterministic challenge](#how-to-calculate-a-deterministic-challenge)) to send to _Client_ and responds with `{ challenge, sdr }` were `sdr` is the [selective disclosure request](#request) defined by the _Service_.

Reviewer: ```suggestion
2. _Service_ creates a random deterministic* _challenge_ to send to _Client_ and responds with `{ challenge, sdr }` were `sdr` is the [selective disclosure request](#request) defined by the _Service_.
```

See comment bellow

3. If `sdr`, _Client_ obtains the information required, and builds the [selective disclosure response](#response)
4. _Client_ builds a JWT with the following payload:
```javascript
{
  iss: `${userDid}`,
  aud: `${serviceUrl}`,
  exp: `${now + 2 min}`,
  nbf: `${now}`,
  iat: `${now}`,
  challenge: `${receivedChallenge}`,
  sdr: `${builtSdr}`
}
```

> and prompts the _User_ to sign it with her `did`'s private key.

Reviewer: Wrong indentation. Breaking `ol`

Author: It is not breaking it

<img width="803" alt="Screen Shot 2020-10-25 at 14 07 10" src="https://user-images.githubusercontent.com/1650329/97113807-6eeb0400-16cb-11eb-889e-733a6a1bffe8.png">

5. _Client_ sends  `POST /signup { response: jwt }` to _Service_

> 6. _Service_ verifies JWT signature, checks if the received `challenge` is ok (by calculating the deterministic one again), and performs business logic over the `sdr`. If it is a valid user, it logges her in by creating an _access token_ and a _refresh token_.

Reviewer: ```suggestion
6. _Service_ verifies JWT signature, checks if the received `challenge` is matches the challenge requested, and performs business logic over the `sdr`. If it is a valid user, it logges her in by creating an _access token_ and a _refresh token_.
```

  - The _access token_ is a JWT signed with the service private key. The JWT MUST have, at least, the following payload:
```javascript
{
  iss: `${serviceDid}`,
  aud: `${serviceUrl}`,
  sub: `${userDid}`
  exp: `${now + 10 min}`, // should be shorter than 15 minutes
  nbf: `${now}`,
  iat: `${now}`,
  // extra information that could be useful for the use case such as user metadata
}

> ```

Reviewer: Same

7. _Client_ authenticates next HTTPS requests using the received _access token_. See [how to send access tokens](#acess-token).
8. If the _access token_ is not expired, _Service_ authorises the request, if not, it answers with an HTTP 401 with the expired token error message.
9. If HTTP 401, _Client_ sends `POST /refresh-token` to _Service_ with the `refresh token`. See [how to send refresh tokens](#refresh-token).
10. _Service_ validates the _refresh token_ and the current session status, if valid, issues new _access token_ (with same data but new expiration), invalidates the received _refresh token_ and issues a new one.
11. _Client_ authenticates next HTTPS requests using the received _access token_.
12. _Service_ authorises the request.

![did auth]({{ site.baseurl }}/assets/img/ssi/10_did_sign_up.png)

The selective disclosure request is optional and it depends on the service needings.
- Open apps: it needs just a proof that the user controls the did. In that case the `challenge` is enough
- Permissioned apps: it needs a proof that the user controls the did and also proofs that the user fullfil the business needings. IE: be older than 18 years old.

> 

Reviewer: ```suggestion

*The challenge can be non-deterministic, but the service will need to store challenge state. See [How to calculate a deterministic challenge](#how-to-calculate-a-deterministic-challenge)

```

A challenge–response authentication is a family of protocols in which one party presents a question ("challenge") and another party must provide a valid answer ("response") to be authenticated. {% include ref.html id="10" %}

> This protocol is based on CHAP{% include ref.html id="11" %} authentication protocol, OAuth 2.0 Authorization Framework{% include ref.html id="13" %} and Decentralized Identifiers (DIDs) v1.0 {% include ref.html id="2" %}.

Reviewer: We are still basing this on SDRs

> 

Reviewer: I would add two sections first, _Motivation_ and _State of the art_. This is kind of introduction to our protocol design and why is important to get the protocol standarized.

It is designed to:

- Authenticate a user by proving than he/she controls a specific did
- Provide an _access token_ to the user that can be reused during a certain amount of time - wallet systems usually request user action to sign messages, enabling reusing access token reduces the amount of signatures required and improves the user experience

> - Allow the user to refresh the _access token_ once it expired - sessions will still be valid beyond the _access token_ expiration time

Reviewer: This is part of the implementation, not design nor motivation. I would say the motivation is to make sessions short to avoid certain attacks

> - Give the user control of the session by providing a logout endpoint that will mark the current session as closed

Reviewer: Is this really important? Does this drive our efforts?

It uses HTTPS as the message transport layer.

Alice needs to access Bob's service, so Bob needs to authenticate Alice:

> 1. Alice sends `POST /request_auth { did }` to Bob, where `did` is Alice's DID

Reviewer: 2 comments on all this section.
1. Let's use _user_ and _app_, is confusing to read _Alice_ and _Bob_ in this case. This nomenclature is better when describing people that are indifferent, _generic_ entities in the protocol. https://en.wikipedia.org/wiki/Alice_and_Bob
2. I would add references to each of the steps in the sequence diagram. Each step should be mapped into an arrow of the sequence diagram.

2. Bob creates a deterministic _challenge_ (see [How to calculate a deterministic challenge](#how-to-calculate-a-deterministic-challenge)) to send to Alice and responds with `{ challenge }`.

> 3. Alice receives Bob's `challenge` and creates a JWT {% include ref.html id="1" %} with the following payload:

Reviewer: Let's add the _registration_ flow as the one and only time you need to disclose your information. SDRs should still be part of this protocol

7. If HTTP 401, Alice sends `POST /refresh-token` to Bob with the `refresh token`. See [how to send refresh tokens](#refresh-token).
8. Bob validates the _refresh token_ and the current session status, if valid, issues new _access token_ (with same data but new expiration), invalidates the received _refresh token_ and issues a new one.
9. Alice authenticates next HTTPS requests using the received _access token_.

![did auth]({{ site.baseurl }}/assets/img/ssi/08_did_auth.png)

## Logout

It will invalidate the current `did` session, so next time `/refresh-token` is invoked, it will not generate a new _access token_.

1. Alices sends `POST /logout` with the current _access token_
2. If the _access token_ is valid, Bob marks the associated _refresh token_ as logged out.
3. Alice sends `POST /refresh-token` to Bob with the `refresh token`
4. Bob does not refresh the _access token_ because the session was closed before.

NOTE: The logout process does not invalidate the current _access token_, it will still be valid until it expires, that's why it matters to implement short validity periods for _access token_. The logout just prevents the _access token_ to be renewed.

> ## How to calculate a deterministic challenge

Reviewer: This can go in an _Appendix_ section. It is not part of the core principles.

Calculate a deterministic challenge prevents the server to maintain a state of emitted challenges while those challenges are valid for a certain amount of time.

Example:

> ```javascript

Reviewer: This is very JS. Using pseudocode for this is always better :)

const challengeExpirationTime = 5 // in minutes
const serverSecret = 'this is the server super secret'
const userDid = `did:ethr:rsk:0x0123456789abcdef`

const timestamp = Math.floor((Date.now() / (challengeExpirationTime * 60 * 1000))

const challenge = keccak256(`${userDid}-${serverSecret}-${timestamp}`)
```

> By doing this, the `challenge` will be valid during maximum 5 minutes. Once the user sends the signed challenge back to the service, the server MUST perform the same calculation and the received challenge must be coincident with the result of that new calculation, if not, an invalid challenge response will be sent.

Reviewer: I would like to read the word `modulo` here. That describes perfectly what this is doing

NOTE: If you decide to use cookies, please make sure that your service is secure enough to prevent Cross Site Request Forgery ([CSRF](https://owasp.org/www-community/attacks/csrf)) atttacks.

### Access token

#### Authorization Header

It must be placed in the `Authorization` header following the `DIDAuth` scheme. This scheme will be present in HTTP Authentication Scheme Registry{% include ref.html id="14" %}.

For example:

```
  GET /resource HTTP/1.1
  Host: server.example.com
  Authorization: DIDAuth my.access.token
```

#### Cookie

When the client performs a `POST /auth`, the server must set the `authorization` cookie with the following attributes: `HttpOnly`, `Secure` and `SameSite=Strict` (this new attribute prevents CSRF, but it is not supported by all the browsers yet). See more information about cookies [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies)

For example:

```
Set-Cookie: authorization=my.access.token; Secure; HttpOnly; SameSite=Strict
```

Then, the client browser will send the cookie on every request.

### Refresh token

#### Body

> It must be placed in the body of the request as a `refresh-token` field.

Reviewer: Why are we using `refresh-token` and not `refresh_token` as we did for `request_auth`? I would follow a unique pattern for all endpoints

For example:

```
  POST /refresh-token HTTP/1.1
  Host: server.example.com

  {
    refresh-token: 'theRefreshToken'
  }
```

#### Cookie

When the client performs a `POST /auth`, the server must set the `refresh-token` cookie with the following attributes: `HttpOnly`, `Secure` and `SameSite=Strict`. 

For example:

```
Set-Cookie: refresh-token=theRefreshToken; Secure; HttpOnly; SameSite=Strict
```

Then, the client browser will send the cookie on every request, so when the client makes a `POST /refresh-token`, the server will replace the existing cookies (`authorization` and `refresh-token`) with new values.

## Implementations

- [`@rskmsart/express-did-auth`](../libraries/express-did-auth) - in progress

## Extensions

> - Separate authentication server from resource server

Reviewer: What is this?

Author: This may be also part of the implementation. The goal is to differentiate the server that keeps the session data from the one that has the resources (API), so this last server could be stateless.

By doing so, the user authenticates and refresh tokens with a Security Token Service that emits tokens can be used in another services that owns the business resources (ie: `/products`)

Reviewer: ```suggestion
- Separate authentication server from resource server. The goal is to differentiate the server that keeps the session data from the one that has the resources (API), so this last server could be stateless. By doing so, the user authenticates and refresh tokens with a Security Token Service that emits tokens can be used in another services that owns the business resources (ie: `/products`)
```

> - Allow the user to associate many dids to the same identity, so the business logic over `did1` and `did2` should be the same if both dids are associated.

Reviewer: ```suggestion
- Allow the user to express control of different identities in a private manner. Lets the user use different identities for a single registration
```
